src/jit/morph.cpp

```diff
@@ -3312,15 +3303,58 @@ GenTreeCall* Compiler::fgMorphArgs(GenTreeCall* callNode)
+                            bool needCpyBlk = false;
+                            if (lclVar != NULL)
+                            {
+                                // If the struct is promoted to registers, it has to be materialized
+                                // on stack. We may want to support promoted structures in
+                                // codegening pugarg_stk instead of creating a copy here.
+                                LclVarDsc*  varDsc = &lvaTable[lclVar->gtLclVarCommon.gtLclNum];
+                                needCpyBlk = varDsc->lvPromoted;
+                            }
+                            else 
+                            {
+                                // The following pattern needs copy to store the
+                                // result of the simd operation. Otherwise, this                               
+                                // pattern reaches codegen that doesn't expect
+                                // addr node.
+                                // Eventually, rationalizer can handle this 
+                                // instead of generating copy here.
+                                //                                
+                                // +--*  obj       simd16
+                                // |  \--*  addr      byref
+                                // |     |  /--*  lclVar    simd16 V05 loc4
+                                // |     \--*  simd      simd16 int -
+                                // |        \--*  lclVar    simd16 V08 tmp1
+                                //
+                                GenTree* addr = argObj->gtOp.gtOp1;
+                                if (addr->OperGet() == GT_ADDR) 
+                                {
+                                    GenTree* addrChild = addr->gtOp.gtOp1;
+                                    if (addrChild->OperGet() == GT_SIMD) 
+                                    {
+                                        needCpyBlk = true;
+                                    }                                    
+                                }
+                            }
                             passStructInRegisters = false;
-                            copyBlkClass = NO_CLASS_HANDLE;
+                            if (needCpyBlk)
+                            {
+                                copyBlkClass = objClass;
+                            }
+                            else
+                            {                                
+                                copyBlkClass = NO_CLASS_HANDLE;
+                            }
                         }
                         else
                         {
                             // The objClass is used to materialize the struct on stack.
                             passStructInRegisters = true;
                             copyBlkClass = objClass;
                         }
+                        
```

Comment:
The following are the struct passing cases on Amd64 Unix that are of interest:

1) struct of size 1,2,4 or 8 in memory/register is getting passed in an arg reg or on stack
2) struct of size 3,5,6,7 in memory/register is getting passed in an arg reg or on stack
1) struct of size 9-16 bytes in memory is getting passed in two arg registers
2) struct of size 9-16 bytes (e.g. SIMD structs Vector3 and Vector4) in a register is getting passed in two arg registers
3) struct of size >16 bytes in memory is getting passed on stack  

Please explain for my benefit for which of the above cases, we will be avoiding copies with these changes?
We need to add TODO comments and open git work items for the ones not currently handled.